database/src/service/read.rs

```diff
@@ -0,0 +1,178 @@
+/// the database result to the caller.
+type ResponseSend = tokio::sync::oneshot::Sender<Response>;
+
+//---------------------------------------------------------------------------------------------------- DatabaseReadHandle
+/// Read handle to the database.
+///
+/// This is cheaply [`Clone`]able handle that
+/// allows `async`hronously reading from the database.
+///
+/// Calling [`tower::Service::call`] with a [`DatabaseReadHandle`] & [`ReadRequest`]
+/// will return an `async`hronous channel that can be `.await`ed upon
+/// to receive the corresponding [`ReadResponse`].
+#[derive(Clone, Debug)]
+pub struct DatabaseReadHandle {
+    /// Sender channel to the database read thread-pool.
+    ///
+    /// We provide the response channel for the thread-pool.
+    pub(super) sender: crossbeam::channel::Sender<(ReadRequest, ResponseSend)>,
+}
+
+impl tower::Service<ReadRequest> for DatabaseReadHandle {
+    type Response = Response; // TODO: This could be a more specific error?
+    type Error = oneshot::error::RecvError; // TODO: always unwrap on channel failure?
+    type Future = ResponseRecv;
+
+    #[inline]
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        todo!()
+    }
+
+    #[inline]
+    fn call(&mut self, _req: ReadRequest) -> Self::Future {
+        todo!()
+    }
+}
+
+//---------------------------------------------------------------------------------------------------- DatabaseReader Impl
+/// Database reader thread.
+///
+/// This struct essentially represents a thread.
+///
+/// Each reader thread is spawned with access to this struct (self).
+pub(super) struct DatabaseReader {
+    /// Receiver side of the database request channel.
+    ///
+    /// Any caller can send some requests to this channel.
+    /// They send them alongside another `Response` channel,
+    /// which we will eventually send to.
+    ///
+    /// We (the database reader thread) are not responsible
+    /// for creating this channel, the caller provides it.
+    receiver: crossbeam::channel::Receiver<(ReadRequest, ResponseSend)>,
+
+    /// TODO: either `Arc` or `&'static` after `Box::leak`
+    /// Access to the database.
+    db: &'static ConcreteEnv,
+}
+
+impl DatabaseReader {
+    /// Initialize the `DatabaseReader` thread-pool.
+    ///
+    /// This spawns `N` amount of `DatabaseReader`'s
+    /// attached to `db` and returns a handle to the pool.
+    ///
+    /// Should be called _once_ per actual database.
+    #[cold]
+    #[inline(never)] // Only called once.
+    pub(super) fn init(db: &'static ConcreteEnv) -> DatabaseReadHandle {
+        // Initalize `Request/Response` channels.
+        let (sender, receiver) = crossbeam::channel::unbounded();
+
+        // TODO: slightly _less_ readers per thread may be more ideal.
+        // We could account for the writer count as well such that
+        // readers + writers == total_thread_count
+        let readers = cuprate_helper::thread::threads().get();
```

Comment:
We don't have to do this now - we can add it later but I would like there to be a config where you would be able to set the amount of threads and other things that can be changed about the DB.